Feature/cdap 13098 provisioner hook up

albertshau opened the pull request.

albertshau committed:
CDAP-13096 refactor program lifecycle to trigger off TMS

Refactoring so that program execution will be triggered off events
on TMS in preparation for additional cluster lifecycle states.
Refactoring the ProgramLifecycleService to remove usage of
ProgramController when possible, as starting a run will first need
to provision the runtime before the ProgramController can be
used.

ProgramLifecycleService will now write a provisioning notification
to TMS to start a program. The ProvisioningService will listen for
it to call the correct provisioner to create the cluster for the
run. Similarly, when a program end state is found, a notification
will be sent on TMS to deprovision the cluster.

Added a 'pending' program run status to indicate the actual
program has not been started. This corresponds to the provisioning
cluster state.

albertshau committed:
CDAP-13098 add a starting program status

With the addition of the provisioning lifecycle, programs will be
in a pending state for minutes until their cluster is provisioned.
It is misleading to categorize this state as 'running', as users
may expect metrics to update and logs to be available, but they
really won't be ready for a few minutes. In order to better reflect
what is happening, added a starting program state instead of just
having running and stopped.

This helps with some unit tests that expect a program to be running
and not starting when the status endpoint returns running.

There is also a bug with stopping programs that are provisioning
but not yet running. This will be addressed in a future PR, but
until then this will fix the tests that fail if they happen to
stop a program while there is no program controller for it yet.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT6383

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/SystemArguments.java:
```diff
@@ -336,6 +343,52 @@ public static Resources getResources(Map<String, String> args, @Nullable Resourc
+  }
+
+  /**
+   * Get the name of the provisioner for the profile
```
Wrong javadoc.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java:
```diff
@@ -777,5 +865,85 @@ private LogLevelUpdater getLogLevelUpdater(RuntimeInfo runtimeInfo) throws Excep
+  @Nullable
+  @Override
+  protected String loadMessageId(DatasetContext datasetContext) throws Exception {
+    ProvisionerDataset.createIfNotExists(datasetFramework);
```
Follow the pattern as in https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java#L113

so that the `datasetContext` is used to get an instance of `ProvisionerDataset`, which is usually cached.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisionerNotificationSubscriberService.java:
```diff
@@ -0,0 +1,151 @@
+ * Subscribes to TMS for program run status changes, calling the ProvisioningService when necessary.
+ */
+public class ProvisionerNotificationSubscriberService extends AbstractNotificationSubscriberService {
+  private static final Logger LOG = LoggerFactory.getLogger(ProgramNotificationSubscriberService.class);
```
Wrong logger.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT6383

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramStateWriter.java:
```diff
@@ -33,7 +33,6 @@
    *
    * @param programRunId the id of the program run
    * @param twillRunId the run id of the twill application
-   * @param artifactId artifact id of the program
```
Why removed while the parameter is still there?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java:
```diff
@@ -162,11 +168,16 @@ private ProgramStatus getExistingAppProgramStatus(ApplicationSpecification appSp
+
+    // A program is starting if there are any PENDING or STARTING run records and no RUNNING run records
+    if (!store.getRuns(programId, ProgramRunStatus.STARTING, 0, Long.MAX_VALUE, 1).isEmpty() ||
+      !store.getRuns(programId, ProgramRunStatus.PENDING, 0, Long.MAX_VALUE, 1).isEmpty()) {
```
These two calls are made with different transactions. I guess we were doing it wrong everywhere. Do you know if we have a JIRA to somehow fix all these?

albertshau replied:
I've opened https://issues.cask.co/browse/CDAP-13296, will add a comment to fix this with the jira number.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -215,14 +240,35 @@ private void writeProgramStatus(ProgramRunId programRunId, ProgramRunStatus prog
+          appMetadataStore.recordProgramDeprovisioning(programRunId, messageIdBytes);
+          appMetadataStore.recordProgramDeprovisioned(programRunId, messageIdBytes);
+        } else {
+          // TODO: remove once runtime monitor emits this message
```
Please add the JIRA CDAP-13295 to this TODO.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -231,32 +277,66 @@ private void writeClusterStatus(ProgramRunId programRunId, ProgramRunClusterStat
     }
   }
 
+  private ProgramOptions getProgramOptions(ProgramId programId, Map<String, String> properties) {
```
createProgramOptions.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java:
```diff
@@ -231,32 +277,66 @@ private void writeClusterStatus(ProgramRunId programRunId, ProgramRunClusterStat
+
+        ProvisionRequest provisionRequest = new ProvisionRequest(programRunId, programOptions, programDescriptor,
+                                                                 userId);
+        provisioningService.provision(provisionRequest);
```
Using provisioingService, which it uses ProvisionerStore, which will start its own tx for each message? I think it should be using the same transaction as the message subscriber in order to have messageId and state changes persisted in the same TX?

Just like we use `AppMetadataStore` directly instead of using `Store` for state updates.

albertshau replied:
Hm right, the initial write to store that a provisioning/deprovisioning operation is happening should be in the same tx, but subsequent writes will need to be in their own tx, as they'll take place minutes later. I can change it to be like that.

albertshau replied:
actually, writing the cluster state in the same TX as the message subscriber is problematic due to the fact that I'm writing state in a separate thread, since we expect that thread to run for minutes. If the cluster state write is done in the same tx as the subscriber, it likely will conflict with any writes done in the thread.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisionerStore.java:
```diff
@@ -0,0 +1,91 @@
+  }
+
+  private ProvisionerDataset getTable(DatasetContext context) throws IOException, DatasetManagementException {
+    ProvisionerDataset.createIfNotExists(datasetFramework);
```
Better follow the https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java#L114

so that it doesn't need to check with `DatasetFramework` for every call.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisionerStore.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * Wrapper around {@link ProvisionerDataset} that runs every operation within a transaction.
+ */
+public class ProvisionerStore {
```
I feel like we should get rid of the `Store` abstraction that automatically wrap everything with transactions. It is error prone as one can easily call two methods without noticing that they are executed in different TX. I think the TX should be controlled explicitly in the service / subscriber.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisioningService.java:
```diff
@@ -29,26 +37,174 @@
     this.provisionerConfigProvider = provisionerConfigProvider;
+    this.provisionerStore = provisionerStore;
+    this.provisionerNotifier = provisionerNotifier;
+    this.executorService = Executors.newCachedThreadPool(
```
Should create this in the `startUp` method. Also, shall we use daemon thread?

albertshau committed:
Fix bugs and address review comments

Fixed a bug where the ProgramDescriptor was not getting stored in
program start event, causing problems when a program is started.

Fixed a bug where programs within a workflow were getting started
twice. Once by the workflow driver and once by the program event
subscriber.

Fixed a race condition in a unit test and fix a unit test
that was failing due the profile always being added as a user
argument.

Fixed a bug where IntegrationTestBase would fail because the
default provisioner does not exist, since there is no extensions
directory to load it from. Changed it so that the default provisioner
is always created in the code instead of loaded as an extension.

Removed ProvisionerStore in favor of ProvisioningService directly
controlling transactions.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java:
```diff
@@ -289,9 +289,19 @@ private ProgramOptions updateProgramOptions(ArtifactId artifactId, ProgramId pro
+      }
+    }
+    if (!options.getArguments().hasOption(ProgramOptionConstants.RUN_ID)) {
+      builder.put(ProgramOptionConstants.RUN_ID, runId.getId());
```
Simpler if you just use normal `Map` and call `putIfAbsent` (available since Java 8).

albertshau committed:
start provisioning tasks after messages have been written in tx

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/SystemArguments.java:
```diff
@@ -336,6 +343,52 @@ public static Resources getResources(Map<String, String> args, @Nullable Resourc
+   * @param profile the profile to add
+   */
+  public static void addProfileArgs(Map<String, String> args, Profile profile) {
+    args.put(PROFILE_NAME, profile.getScopedName());
```
I think it's better to have these two puts after the loop below, otherwise user can override these two.

chtyim replied:
Also, what happen if a provisioner actually has a property called "system.profile.name"? Shall we store the profile name and profile provision in the system arguments instead of the user arguments?

albertshau replied:
profile properties are prefixed with 'system.profile.properties.', so it would become 'system.profile.properties.system.profile.name', and the user would have to use that key in order to override it. I think having it in user arguments is consistent with other things like the resources and tx level, and having it there also indicates that it can be overridden by the user.

albertshau replied:
hm though I guess we don't add those values to user arguments, we just allow them to be overridden from them. Though we do always add 'logical.start.time'. I guess I can move them to system arguments.

albertshau committed:
start the provisioning service so it loads provisioners

albertshau committed:
store profile properties in system args instead of user args
